connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala

```diff
@@ -664,7 +665,53 @@ class SparkConnectPlanner(val session: SparkSession) {
         input: proto.Relation,
         groupingExprs: java.util.List[proto.Expression],
         sortingExprs: java.util.List[proto.Expression]): UntypedKeyValueGroupedDataset = {
-      val logicalPlan = transformRelation(input)
+      apply(transformRelation(input), groupingExprs, sortingExprs)
+    }
+
+    private def apply(
+        logicalPlan: LogicalPlan,
+        groupingExprs: java.util.List[proto.Expression],
+        sortingExprs: java.util.List[proto.Expression]): UntypedKeyValueGroupedDataset = {
+      if (groupingExprs.size() == 1) {
+        createFromGroupByKeyFunc(logicalPlan, groupingExprs, sortingExprs)
+      } else if (groupingExprs.size() > 1) {
```

Comment:
This feels a bit dirty. I understand that the handling the grouping function seems a bit different, but I wonder if we need the special single function result flattening behavior in the grand scheme of things, especially when you consider we are going to end up with a tupled dataset anyway.

Reply:
I do not see a common path here. The nasty part of this is we hide a logic inside the grouping_exprs using the count of the expressions. The alternative I can think is a UnresolvedFunc or new Expression which allow us to add more logics. e.g.
```
message KeyValueGroupedDataset { // New Expression or Unresolved Func
 // (Required) Input user-defined function. Defines the grouping func
 CommonInlineUserDefinedFunction grouping_func = 1;

 // (Optional) Extra grouping expressions needed for RelationalGroupedDataset
 repeat Expression grouping_expressions = 2;
}
```